Sources/TuistCache/ContentHashing/ContentHasher.swift

```diff
@@ -0,0 +1,49 @@
+import Foundation
+import Basic
+import TuistCore
+import TuistSupport
+
+public protocol FileContentHashing {
+    func hash(fileAtPath: AbsolutePath) throws -> String
+}
+
+public protocol ContentHashing: FileContentHashing {
+    func hash(_ string: String) throws -> String
+    func hash(_ strings: [String]) throws -> String
+}
+
+/// ContentHasher
+/// The single source of truth for hashing content
+/// Using md5 checksum to uniquely hash strings and data
+/// Consider using CacheContentHasher to avoid recalculating the same hash twice
+
+public final class ContentHasher: ContentHashing {
+    private let fileHandler: FileHandling
+
+    public init(fileHandler: FileHandling = FileHandler.shared) {
+        self.fileHandler = fileHandler
+    }
+
+    // MARK: - ContentHashing
+
+    public func hash(_ string: String) throws -> String {
+        guard let hash = string.checksum(algorithm: .md5) else {
+            throw ContentHashingError.stringHashingFailed(string)
+        }
+        return hash
+    }
+
+    public func hash(_ strings: [String]) throws -> String {
+        return try hash(strings.joined())
+    }
+
+    public func hash(fileAtPath filePath: AbsolutePath) throws -> String {
+        guard let sourceData = try? fileHandler.readFile(filePath) else {
```

Comment:
Reading a file and file not found are 2 different things. I propose changing these lines to:

```swift
guard let fileHandler.exists(filePath) else {
  throw ContentHashingError.fileNotFound(filePath)
}
let data = try fileHandler.readFile(filePath)
```
Have you compared your current approach, reading into memory and hashing, with running the md5 command in the system? *(in terms of performance)*.

Reply:
Updated with the suggestion. I'd defer performance optimizations once we have a working implementation

Reply:
Fair enough